RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc

```diff
@@ -0,0 +1,586 @@
+   // If the template recontruction fails, we want to return 1.0 for now
+   theClusterParam.templProbY_ = theClusterParam.templProbX_ = theClusterParam.templProbQ_ = 1.0f;
+   theClusterParam.templQbin_ = 0;
+   // We have a boolean denoting whether the reco failed or not
+   theClusterParam.hasFilledProb_ = false;
+   
+   // ******************************************************************
+   //--- Call 2D TemplateReco
+   //
+   float locBz = theDetParam.bz;
+   float locBx = theDetParam.bx;
+
+   //--- Input:
+   //   edgeflagy - (input) flag to indicate the present of edges in y: 
+   //           0=none (or interior gap),1=edge at small y, 2=edge at large y, 3=edge at either end
+   //
+   //   edgeflagx - (input) flag to indicate the present of edges in x: 
+   //           0=none, 1=edge at small x, 2=edge at large x
+   //
+   //   These two variables are calculated in setTheClu() and stored in edgeTypeX_ and edgeTypeY_
+   //
+   //--- Output:
+   //   deltay - (output) template y-length - cluster length [when > 0, possibly missing end]
+   //   npixels - ???     &&& Ask Morris
+
+   float edgeTypeY = theClusterParam.edgeTypeY_ ;  // the default, from PixelCPEBase
+   if ( theClusterParam.recommended3D_ ) {
+     //  Cluster is not on edge, but instead the normal TemplateReco discovered that it is
+     //  shorter than expected.  So let the 3D algorithm try extending it on both sides, in case
+     //  there is a dead double-column on either side.  (We don't know which.)
+     edgeTypeY = 3;
+   }
+
+   float deltay = 0;    // return param
+   int npixels = 0;     // return param
+
+   theClusterParam.ierr2 =
+   PixelTempReco3D( ID, theClusterParam.cotalpha, theClusterParam.cotbeta,
+                   locBz, locBx,
+		   edgeTypeY , theClusterParam.edgeTypeX_ ,
+                   clusterPayload,
+                   templ2d,
+                   theClusterParam.templYrec_, theClusterParam.templSigmaY_, 
+                   theClusterParam.templXrec_, theClusterParam.templSigmaX_, 
+		   theClusterParam.templProbXY_,
+         	   theClusterParam.probabilityQ_,
+		   theClusterParam.qBin_,
+		   deltay, npixels
+                   );
+   // ******************************************************************
+   
+   //--- Check exit status
+   if unlikely( theClusterParam.ierr2 != 0 )
+   {
+      LogDebug("PixelCPEClusterRepair::localPosition") <<
+      "3D reconstruction failed with error " << theClusterParam.ierr2 << "\n";
+      
+      // GG: what do we do in this case?  For now, just return the cluster center of gravity in microns
+      // In the x case, apply a rough Lorentz drift average correction
+      // To do: call PixelCPEGeneric whenever PixelTempReco2D fails
+      float lorentz_drift = -999.9;
+      if ( ! GeomDetEnumerators::isEndcap(theDetParam.thePart) )
+         lorentz_drift = 60.0f; // in microns
```

Comment:
this is used multiple times. Define a named constant instead

Reply:
This handles the case where the 1D template fit failed.  This almost never happens.  However, you are right that we should use a better number, but I should probably get it from the DB -- the number exists in CPE Generic.  But I need to think about all this.  I added a note in the comments that we should fix this.